### PR TITLE
Supported for extending the debug property pages

### PR DESCRIPTION
**Pull request by BillHiebert**
The high level design notes are:
1. The list of commands which appear in the Launch dropdown are provided by the set of ILaunchSettingsUIProvider's. There are implementations for Project and executable in the base project system to support Console\ClassLibraries
image

2. The providers can optionally add Custom UI to the dialog to make it easy to set custom values. This UI appears after the default values. 
3. Default properties can be hidden if they don't apply using the EnableProperty method
4. Sequence
    a. Dialog loads the providers to build the dropdown list. If there are commands in the current launch settings for which there isn't a provider installed, the command will be shown when that profile is active, but other profiles can't be converted to it. All fields are disabled and it becomes R/O. Note we should allow deletion.
    b. When the profile is changed (or set for the first time), the ProfileSelected() method will be called on the provider which matches the command and the current set of profiles is passed to it.  Changes made to settings need to be applied to this set of profiles. 
    c. If the provider has a control it will be inserted into the dialog and resized to match the existing layout. In general the control should match the 3 column style of the main control.
    d. When the provider changes, the debug view model listens to property changes its controls control datacontext (assuming it has a custom control, it may not) so that changes to the control affect the dirty state of the dialog. When the active provider changes it disconnects from the current provider and connects to the new one.
    e. Similarly, for validation, the control  hooks into INotifyDataErrorInfo of the controls datacontext. If there are errors in the control, the profile dropdown is disabled until the errors are corrected (this is what is done for env variables already). 
    f. If there are validation errors, save will fail with an error message informing the user they need to correct the errors before  save can continue.
5. Writable versions of ILaunchSettings\ILaunchProfile are used to hold settings in the dialog. This ensures the current immutable snapshot is not perturbed while the editing is going on. 

Unit tests still need to be completed

@jinujoseph @srivatsn @abpiskunov

### CODE REVIEW

**Comment by BillHiebert** [comment]
@mlorbetske

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj [hunk 1]
```diff
@@ -235,8 +237,9 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Include="ProjectSystem\VS\PropertyPages\PropertyPageResources.resx">
-      <Generator>ResXFileCodeGenerator</Generator>
+      <Generator>PublicResXFileCodeGenerator</Generator>
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj]: Why the change to public generation?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageControl.xaml.cs [hunk 2]
```diff
@@ -70,7 +76,59 @@ private void ClearGridError(DataGrid dataGrid)
...
+        
+        /// <summary>
+        /// Called when a property changes on the view model. Used to detect when the custom UI changes so that
+        /// the size of its grids can be determined
+        /// </summary>
+        private void ViewModel_PropertyChanged(Object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            if(e.PropertyName.Equals(nameof(DebugPageViewModel.ActiveProviderUserControl)))
+            {
+                _customControlLayoutUpdateRequired = true;
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageControl.xaml.cs]: I'm not that familiar with XAML, so I'm assuming that these event handlers are always on the UI thread?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 3]
```diff
@@ -38,53 +38,29 @@ private IProjectThreadingService ProjectThreadingService
...
+
         public event EventHandler ClearEnvironmentVariablesGridError;
         public event EventHandler FocusEnvironmentVariablesGridRow;
+
+        public event EventHandler<DataErrorsChangedEventArgs> ErrorsChanged;
+
         public DebugPageViewModel()
         {
-
+
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Whitespace.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 4]
```diff
@@ -127,13 +104,68 @@ public LaunchType SelectedLaunchType
                             ExecutablePath = null;
                         }
 
-                        OnPropertyChanged(nameof(IsExecutable));
-                        OnPropertyChanged(nameof(IsProject));
+                        // Let the active provider know of the changes. 
+                        SwitchProviders(oldActiveProvider);
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Where is ActiveProvider updated here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 5]
```diff
@@ -127,13 +104,68 @@ public LaunchType SelectedLaunchType
...
+                }
+            }
+
+            // Now hook into the current providers notifications. We do that after having set the profile on the provider
+            // so that we don't get notifications while the control is initializing. Note that this is likely the first time the 
+            // custom control is asked for and we want to call it and have it created prior to setting the active profile
+            var customControl = ActiveProvider?.CustomUI;
+            if(customControl != null)
+            {
+                ActiveProvider?.ProfileSelected(CurrentLaunchSettings);
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: No need for `?.`, ActiveProvider is guaranteed to be non-null here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 6]
```diff
@@ -127,13 +104,68 @@ public LaunchType SelectedLaunchType
...
+
+                context = customControl.DataContext as INotifyPropertyChanged;
+                if(context != null)
+                {
+                    context.PropertyChanged += OnCustomUIStateChanged;
+                }
+
+                if(context is INotifyDataErrorInfo)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: You can introduce a variable with this `is` to avoid the cast.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 7]
```diff
@@ -239,32 +271,40 @@ public bool HasLaunchOption
...
-        public bool IsExecutable
+        public bool SupportsExecutable
         {
-            get
-            {
-                if (SelectedLaunchType == null)
-                {
-                    return false;
-                }
+            get {  return ActiveProviderSupportsProperty(UIProfilePropertyName.Executable); }
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: We can make these more idiomatic by using `=>` here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 8]
```diff
@@ -239,32 +271,40 @@ public bool HasLaunchOption
...
         }
-        
+
+        public bool SupportsEnvironmentVariables
+        {
+            get {  return ActiveProviderSupportsProperty(UIProfilePropertyName.EnvironmentVariables); }
+        }
+
+        /// <summary>
+        /// Helper returns true if there is an actife provider and it supports the specified property
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Spelling on active.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 9]
```diff
@@ -239,32 +271,40 @@ public bool HasLaunchOption
...
+        /// </summary>
+        private bool ActiveProviderSupportsProperty(string propertyName)
+        {
+            var activeProvider = ActiveProvider;;
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Extra semicolon

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 7 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Regardless of the style you settle on, all of the properties in the class should have the same formatting. The ones below have everything on a newline.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj (hunk 1 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Microsoft.VisualStudio.ProjectSystem.Managed.VS.csproj]: The XAML requires it since it has static bindings. The only reason the code worked before is someone had hand edited the generated CS file to make the class and members public. As soon as I made a resx file change, the dialog no longer worked.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageControl.xaml.cs (hunk 2 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageControl.xaml.cs]: Correct.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 4 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: It is calculated from the change in the launch type.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 3 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 6 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 7 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: I made them match the other properties

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 8 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 9 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 10]
```diff
@@ -455,38 +497,45 @@ private void UpdateActiveProfile()
...
         {
             ILaunchSettingsProvider provider = GetDebugProfileProvider();
-            if (EnvironmentVariables != null && EnvironmentVariables.Count > 0)
+            if (EnvironmentVariables != null && EnvironmentVariables.Count > 0 && SelectedDebugProfile != null)
             {
-                SelectedDebugProfile.MutableEnvironmentVariables = EnvironmentVariables.CreateDictionary();
+                SelectedDebugProfile.EnvironmentVariables.Clear();
+                foreach(var kvp in EnvironmentVariables)
+                {
+                    SelectedDebugProfile.EnvironmentVariables.Add(kvp.Name, kvp.Value);
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Is it possible for the SelectedDebugProfile to change while this async task is running?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 11]
```diff
@@ -506,74 +555,56 @@ private void EnvironmentVariables_ValidationStatusChanged(object sender, EventAr
...
-                    var debugProfiles = new ObservableCollection<LaunchProfile>();
+                // Update the set of settings and generate a property change so the list of profiles gets updated. Note that we always
+                // clear the active profile on the CurrentLaunchSettings so that when we do set one and property changed event is set
+                CurrentLaunchSettings = newSettings;
+                CurrentLaunchSettings.ActiveProfile = null;
+                NotifyProfileCollectionChanged();
 
-                    foreach (var profile in profiles.Profiles)
+                // If we have a selection, we want to leave it as is
+                if (curProfileName == null || newSettings.Profiles.FirstOrDefault(p => { return LaunchProfile.IsSameProfileName(p.Name, curProfileName); }) == null)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: This line can be significantly condensed by doing `FirstOrDefault(p => LaunchProfile.IsSameProfileName(p.Name, curProfileName))`, getting rid of the unnecessary brackets, return, and semicolon.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 12]
```diff
@@ -771,26 +842,28 @@ public ICommand DeleteProfileCommand
...
         {
-            var profile = new LaunchProfile() { Name = name, CommandName = commandName };
-            DebugProfiles.Add(profile);
+            var profile = new WritableLaunchProfile() { Name = name, CommandName = commandName };
+            CurrentLaunchSettings.Profiles.Add(profile);
 
-            // Fire a property changed so we can get the page to be dirty when we add a new profile
+            NotifyProfileCollectionChanged();
+            
+            // Todo: Do we need this extra property change
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Not sure, but we should have a bug for it to track.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 13]
```diff
@@ -864,13 +951,61 @@ public override void ViewModelDetached()
...
+        /// We are considered in error if we have errors, or the currently active control has errors
+        /// </summary>
+        public bool HasErrors
+        {
+            get
+            {
+                var notifyDataError = ActiveProvider?.CustomUI?.DataContext as INotifyDataErrorInfo;
+                if(notifyDataError != null)
+                {
+                    if(notifyDataError.HasErrors)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Just return HasErrors.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 14]
```diff
@@ -864,13 +951,61 @@ public override void ViewModelDetached()
...
+
+        /// <summary>
+        /// We are considered in error if we have errors, or the currently active control has errors
+        /// </summary>
+        public bool HasErrors
+        {
+            get
+            {
+                var notifyDataError = ActiveProvider?.CustomUI?.DataContext as INotifyDataErrorInfo;
+                if(notifyDataError != null)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Can be simplified with C# 7

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs [hunk 15]
```diff
@@ -1,23 +1,23 @@
...
+using System.Windows.Controls;
 using System.Windows.Input;
 using Microsoft.VisualStudio.ProjectSystem.Debug;
 using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
 using Microsoft.VisualStudio.Shell;
 
 namespace Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages
 {
-    internal class DebugPageViewModel : PropertyPageViewModel
+    internal class DebugPageViewModel : PropertyPageViewModel, INotifyDataErrorInfo
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: This class is big enough that we should file a bug to refactor components out. There's over 1000 lines of code here and it's pretty hard to see how all the pieces fit together.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Windows.Controls;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages
+{
+    /// <summary>
+    /// Interface definition which allows a launch settings provider to participate in the debug property page UI. The Set of
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: This isn't an interface.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    /// Interface definition which allows a launch settings provider to participate in the debug property page UI. The Set of 
+    /// LaunchSettingUIProviders provides the set of entries for the debug dropdown command list.
+    /// </summary>
+    [Export( typeof(ILaunchSettingsUIProvider))]
+    [AppliesTo(ProjectCapability.LaunchProfiles)]
+    [Order(0)]              // Lowest priority to llow this to be overridden
+    internal class ExecutableLaunchSettingsUIProvider : ILaunchSettingsUIProvider
+    {
+        [ImportingConstructor]
+        public ExecutableLaunchSettingsUIProvider(UnconfiguredProject uncProject)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: Is this just imported to control the MEF scope? If so, could you add a comment?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        {
+
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: Extra newline.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs (hunk 16 above)
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: In fact, all of these doc comments should probably be on the interface, not the class.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ProjectLaunchSettingsUIProvider.cs [hunk 19]
```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Windows.Controls;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PropertyPages
+{
+    /// <summary>
+    /// Interface definition which allows a launch settings provider to participate in the debug property page UI. The Set of
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ProjectLaunchSettingsUIProvider.cs]: Same comment about the comments.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPageResources.resx [hunk 20]
```diff
@@ -117,21 +117,12 @@
   <resheader name="writer">
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
-  <data name="AdminRequiredForPort" xml:space="preserve">
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/PropertyPageResources.resx]: /cc @nguerrera for loc changes.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs [hunk 21]
```diff
@@ -37,6 +39,34 @@ public LaunchSettings(LaunchSettingsData settingsData, string activeProfile = nu
...
+            {
+                Profiles = Profiles.Add(new LaunchProfile(profile));
+            }
+            
+            // For global settings we want to make new copies of each entry so that the snapshot remains immutable. If the object implements 
+            // ICloneable that is used, otherwise, it is serialized back to json, and a new object rehydrated from that
+            GlobalSettings = ImmutableDictionary<string, object>.Empty;
+            foreach(var kvp in settings.GlobalSettings)
+            {
+                if(kvp.Value is ICloneable)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs]: C#7 to avoid casting.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs [hunk 22]
```diff
@@ -37,6 +39,34 @@ public LaunchSettings(LaunchSettingsData settingsData, string activeProfile = nu
...
+                {
+                    GlobalSettings = GlobalSettings.Add(kvp.Key, ((ICloneable)kvp.Value).Clone());
+                }
+                else
+                {
+                    string jsonString = JsonConvert.SerializeObject(kvp.Value, Formatting.Indented, new JsonSerializerSettings() { NullValueHandling = NullValueHandling.Ignore});
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs]: If no one else is using this string, why do we care what Formatting is used? And in general, is there really no better way to clone an object than to turn it into a string and then parse the string?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+            if(profile.OtherSettings  != null)
+            {
+                OtherSettings = new Dictionary<string, object>(profile.OtherSettings, StringComparer.Ordinal);
+            }
+        }
+
+        /// <summary>
+        /// IWritableDebug profile members
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs]: This isn't really a doc comment, is it? Or if it is, it applies to all the members, not just the top ones.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 10 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: No. Save is actually wrapped in Synchronous call blocking the UI thread.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+        /// </summary>
+        public static bool ProfilesAreEqual(IWritableLaunchProfile debugProfile1, IWritableLaunchProfile debugProfile2)
+        {
+            // Same instance are equal
+            if (debugProfile1 == debugProfile2)
+            {
+                return true;
+            }
+
+            if (!string.Equals(debugProfile1.Name, debugProfile2.Name, StringComparison.Ordinal) ||
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs]: Just return this comparison.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 11 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Yes. I've fixed a few of these

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchSettings.cs [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+                }
+            }
+            
+            // For global settings we want to make new copies of each entry so that the snapshot remains immutable. If the object implements 
+            // ICloneable that is used, otherwise, it is serialized back to json, and a new object rehydrated from that
+            if(settings.GlobalSettings != null)
+            {    
+                foreach(var kvp in settings.GlobalSettings)
+                {
+                    if(kvp.Value is ICloneable)
```
Comment (333fred) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchSettings.cs]: Same comment about avoiding the cast

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 12 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: I put back the original comment. Not sure why I added the todo in the first place.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 13 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: My original thought was that if we add errors (like convert the way env variable errors hare handled) to the main dialog, it would do its check later where I currently return false. But I fixed this as you suggested.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs (hunk 14 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/DebugPageViewModel.cs]: Yes. Made that change too.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs (hunk 17 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: Yes. fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs (hunk 16 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: Fixed the comment. Doc comments are already on the interface.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ProjectLaunchSettingsUIProvider.cs (hunk 19 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ProjectLaunchSettingsUIProvider.cs]: fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs (hunk 18 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PropertyPages/ExecutableLaunchSettingsUIProvider.cs]: fixed

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs (hunk 22 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs]: I couldn't come up with a better one that would do a proper deep copy. The formatting matters because it will create an empty value which may become non-null in the clone. When it is later persisted, the empty entry will be created in the launch settings file.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs (hunk 21 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/LaunchSettings.cs]: done

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs (hunk 23 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs]: I removed the comment

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs (hunk 24 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchProfile.cs]: I left this one as I think it is easier to read

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchSettings.cs (hunk 25 above)
Comment (BillHiebert) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Debug/WritableLaunchSettings.cs]: Done.
